controller/contour/deployment.go

```diff
@@ -0,0 +1,423 @@
+	deployCfgMapName = "contour"
+	// deployNsEnvVar is the name of the contour namespace environment variable.
+	deployNsEnvVar = "CONTOUR_NAMESPACE"
+	// deployPodEnvVar is the name of the contour pod name environment variable.
+	deployPodEnvVar = "POD_NAME"
+	// certsVolumeName is the name of the contour certificates volume.
+	certsVolumeName = "contourcert"
+	// certsVolumeMountDir is the directory name of the contour certificates volume.
+	certsVolumeMountDir = "certs"
+	// certsSecretName is the name of the secret used as the certificate volume source.
+	certsSecretName = certsVolumeName
+	// cfgVolumeName is the name of the contour configuration volume.
+	cfgVolumeName = "contour-config"
+	// cfgVolumeMountDir is the directory name of the contour configuration volume.
+	cfgVolumeMountDir = "config"
+	// cfgFileName is the name of the contour configuration file.
+	cfgFileName = "contour.yaml"
+)
+
+// ensureDeployment ensures a deployment exists for the given contour.
+func (r *Reconciler) ensureDeployment(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	haveDepl, current, err := r.currentDeployment(ctx, contour)
+	if err != nil {
+		return err
+	}
+
+	version, err := stringsutil.VersionFromImage(deployImageName)
+	if err != nil {
+		return fmt.Errorf("failed to parse version from image %s: %v", deployImageName, err)
+	}
+	desired, err := desiredDeployment(contour, version)
+	if err != nil {
+		return fmt.Errorf("failed to build deployment: %v", err)
+	}
+
+	switch {
+	case !haveDepl:
+		if err := r.createDeployment(ctx, desired); err != nil {
+			return err
+		}
+	case haveDepl:
+		if _, err := r.updateDeployment(ctx, current, desired); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ensureDeploymentDeleted ensures the deployment for the provided contour
+// is deleted.
+func (r *Reconciler) ensureDeploymentDeleted(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	key := types.NamespacedName{
+		Namespace: contour.Spec.Namespace.Name,
+		Name:      contour.Name,
+	}
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: key.Namespace,
+			Name:      key.Name,
+		},
+	}
+
+	if err := r.Client.Delete(ctx, deployment); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	r.Log.Info("deleted deployment", "namespace", deployment.Namespace, "name", deployment.Name)
+	return nil
+}
+
+// desiredDeployment returns the desired deployment for the provided contour
+// using version as the "app.kubernetes.io/version" label value.
+func desiredDeployment(contour *operatorv1alpha1.Contour, version string) (*appsv1.Deployment, error) {
+	labels := map[string]string{
+		"app.kubernetes.io/name":       "contour",
+		"app.kubernetes.io/instance":   contour.Name,
+		"app.kubernetes.io/version":    version,
+		"app.kubernetes.io/component":  "ingress-controller",
+		"app.kubernetes.io/part-of":    "project-contour",
```

Comment:
`part-of` has some interpretation. I'd suggest `projectcontour.io` to refer to Contour the project, or maybe the name of the `Contour` object that is owning this controller?

Reply:
We already have "app.kubernetes.io/instance" and ""contour.operator.projectcontour.io/owning-contour" that refer to the Contour name. I'll remove this label for now until the community settles on a spec.